raiden-ts/src/transport/epics/presence.ts

```diff
@@ -161,6 +161,68 @@ type MatrixPresenceEvent = Omit<MatrixEvent, 'getContent'> & {
   getContent: () => PresenceContent;
 };
 
+// observable of all addresses whose presence monitoring was requested since init
+function toMonitor$(action$: Observable<RaidenAction>) {
+  const toMonitor = new Set<Address>();
+  return action$.pipe(
+    filter(matrixPresence.request.is),
+    scan((toMonitor, request) => toMonitor.add(request.meta.address), toMonitor),
+    startWith(toMonitor),
+  );
+}
+
+function fetchPresence$(
+  user: User,
+  address: Address,
+  event: MatrixPresenceEvent,
+  {
+    log,
+    config$,
+    latest$,
+    matrix$,
+  }: Pick<RaidenEpicDeps, 'log' | 'config$' | 'latest$' | 'matrix$'>,
+) {
+  return matrix$.pipe(
+    switchMap((matrix) =>
+      combineLatest([
+        // always fetch profile info, to get up-to-date displayname, avatar_url & presence
+        defer(async () => matrix.getProfileInfo(user.userId)),
+        defer(async () =>
+          AVAILABLE.includes(event.getContent().presence)
+            ? event.getContent()
+            : getUserPresence(matrix, user.userId),
+        ),
+        latest$.pipe(pluckDistinct('rtc', address)),
+      ]),
+    ),
+    map(([profile, { presence, last_active_ago }, rtc]) => {
+      // errors raised here will be logged and ignored on catchError below
+      assert(profile?.displayname, 'no displayname');
+      // ecrecover address, validating displayName is the signature of the userId
+      const recovered = verifyMessage(user.userId, profile.displayname) as Address | undefined;
+      assert(recovered === address, `invalid displayname signature: ${recovered} !== ${address}`);
+      if (presence !== user.presence)
+        log.warn('Presence mismatch', { user, presence, last_active_ago });
+      // react on both presence events and rtc channel presence
+      const available = AVAILABLE.includes(presence) || !!rtc;
+      return matrixPresence.success(
+        {
+          userId: user.userId,
+          available,
+          ts: last_active_ago ? Date.now() - last_active_ago : user.lastPresenceTs ?? Date.now(),
+          caps: parseCaps(profile.avatar_url),
+        },
+        { address: recovered },
+      );
+    }),
+    retryWhile(intervalFromConfig(config$), { onErrors: [429] }),
```

Comment:
Should we include 500 here as well?

Reply:
I don't think 500 is needed here, because there's a `catchError` right below. If server misbehaves, it's fine to handle and skip a single presence update.